### PR TITLE
registry: Refactor requestfactory to use http.RoundTrippers

### PR DESCRIPTION
**Pull request by tiborvass**
**Replaced by #13265**
# 

This patch removes the need for requestFactories and decorators
by implementing http.RoundTripper transports instead.

It refactors some challenging-to-read code.

NewSession now takes an *http.Client that can already have a
custom Transport, it will add its own auth transport by wrapping
it.

The idea is that callers of http.Client should not bother
setting custom headers for every handler but instead it should
be transparent to the callers of a same context.

This patch is needed for future refactorings of registry,
namely refactoring of the v1 client code.

Signed-off-by: Tibor Vass tibor@docker.com

### CODE REVIEW

**Comment by tiborvass** [comment]
Ping @dmcgowan @stevvooe

**Comment by tiborvass** [comment]
Failing tests...

File: registry/registry.go [hunk 1]
```diff
@@ -31,65 +35,17 @@ const (
...
-		Transport:     httpTransport,
-		CheckRedirect: AddRequiredHeadersToRedirectedRequests,
-		Jar:           jar,
-	}
+type httpsTransport struct {
+	*http.Transport
 }
 
-func doRequest(req *http.Request, jar http.CookieJar, timeout TimeoutType, secure bool) (*http.Response, *http.Client, error) {
+func (tr *httpsTransport) RoundTrip(req *http.Request) (*http.Response, error) {
```
Comment (tiborvass) [inline registry/registry.go]: If someone wonders why do we set tlsconfig in a roundtrip, it's because it's so as to match the current behavior in master: we generate the certpool on every-goddam-request. It's not great, but it allows people to just put the certs in /etc/docker/certs.d/.../ and let docker "pick it up" immediately. Would prefer an fsnotify implementation, but that's not the scope of this PR

Comment (diogomonica) [inline registry/registry.go]: wow, lame.

File: registry/registry.go [hunk 2]
```diff
@@ -136,24 +92,135 @@ func doRequest(req *http.Request, jar http.CookieJar, timeout TimeoutType, secur
...
 				if !hasFile(fs, certName) {
-					return nil, nil, fmt.Errorf("Missing certificate %s for key %s", certName, keyName)
+					return nil, fmt.Errorf("Missing certificate %s for key %s", certName, keyName)
 				}
 			}
 		}
+		if tr.Transport.TLSClientConfig == nil {
+			tr.Transport.TLSClientConfig = &tls.Config{
+				// Avoid fallback to SSL protocols < TLS1.0
+				MinVersion: tls.VersionTLS10,
```
Comment (diogomonica) [inline registry/registry.go]:
```go
config := tls.Config{
    MinVersion:               tls.VersionTLS10,
    PreferServerCipherSuites: true,
    CipherSuites: []uint16{
        tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
        tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
        tls.TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA,
        tls.TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,
        tls.TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,
        tls.TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA,
        tls.TLS_RSA_WITH_AES_128_CBC_SHA,
        tls.TLS_RSA_WITH_AES_256_CBC_SHA},
}
```

File: registry/registry.go [hunk 3]
```diff
@@ -136,24 +92,135 @@ func doRequest(req *http.Request, jar http.CookieJar, timeout TimeoutType, secur
...
+			}
+		}
+		tr.Transport.TLSClientConfig.RootCAs = roots
+		tr.Transport.TLSClientConfig.Certificates = certs
+	}
+	return tr.Transport.RoundTrip(req)
+}
+
+func DefaultTransport(timeout TimeoutType, secure bool) http.RoundTripper {
+	tlsConfig := tls.Config{
```
Comment (diogomonica) [inline registry/registry.go]: See the `tls.Config` I proposed above this.

File: registry/registry.go [hunk 4]
```diff
@@ -136,24 +92,135 @@ func doRequest(req *http.Request, jar http.CookieJar, timeout TimeoutType, secur
...
+		// Avoid fallback to SSL protocols < TLS1.0
+		MinVersion:         tls.VersionTLS10,
+		InsecureSkipVerify: !secure,
```
Comment (diogomonica) [inline registry/registry.go]: Could use a comment saying what the default value of secure is

File: registry/registry.go [hunk 5]
```diff
@@ -136,24 +92,135 @@ func doRequest(req *http.Request, jar http.CookieJar, timeout TimeoutType, secur
...
+	for k, v := range tr.Headers {
+		req.Header[k] = v
+	}
+	return tr.RoundTripper.RoundTrip(req)
+}
+
+type debugTransport struct{ http.RoundTripper }
+
+func (tr debugTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	fmt.Println(req.Method, req.URL)
```
Comment (diogomonica) [inline registry/registry.go]: Can turn these three into one fmt.Printf

File: registry/session.go [hunk 6]
```diff
@@ -20,64 +21,96 @@ import (
...
 )
 
 type Session struct {
-	authConfig    *cliconfig.AuthConfig
-	reqFactory    *requestdecorator.RequestFactory
 	indexEndpoint *Endpoint
-	jar           *cookiejar.Jar
-	timeout       TimeoutType
+	client        *http.Client
+	// TODO: remove authConfig
```
Comment (diogomonica) [inline registry/session.go]: format of TODO's should be: `(diogo): Remove authConfig`

File: registry/session.go [hunk 7]
```diff
@@ -20,64 +21,96 @@ import (
...
+	token              []string
+}
 
-	if timeout {
-		r.timeout = ReceiveTimeout
+func (tr *authTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	req = cloneRequest(req)
+
+	if tr.alwaysSetBasicAuth {
+		//debug:
```
Comment (diogomonica) [inline registry/session.go]: Why is this here.

File: registry/registry.go (hunk 2 above)
Comment (tiborvass) [inline registry/registry.go]: thanks i'll take care of that in another PR if you don't mind

File: registry/registry.go (hunk 4 above)
Comment (tiborvass) [inline registry/registry.go]: secure depends on daemon options, please see https://github.com/docker/docker/blob/e00cfbb6f2b953301bfcd7270e9bdfde1a2d17c7/registry/config.go#L128

File: registry/registry.go (hunk 1 above)
Comment (tiborvass) [inline registry/registry.go]: indeed

File: registry/registry.go (hunk 5 above)
Comment (tiborvass) [inline registry/registry.go]: used httputil.DumpRequestOut instead

File: registry/session.go (hunk 7 above)
Comment (tiborvass) [inline registry/session.go]: so that you notice it :P removed it thanks!

**Comment by tiborvass** [comment]
@diogomonica the newer PR is at https://github.com/docker/docker/pull/13265 thanks for review though!
